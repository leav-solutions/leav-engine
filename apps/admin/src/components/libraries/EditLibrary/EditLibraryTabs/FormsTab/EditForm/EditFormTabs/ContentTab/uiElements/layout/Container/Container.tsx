// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import React, {useRef} from 'react';
import {useDrop, XYCoord} from 'react-dnd';
import {useTranslation} from 'react-i18next';
import styled from 'styled-components';
import {
    defaultContainerId,
    FormBuilderActionTypes,
    IFormBuilderStateAndDispatch
} from '../../../formBuilderReducer/formBuilderReducer';
import FormElementWrapper from '../../../FormLayout/FormLayoutElementWrapper/FormElementWrapper';
import {DraggableElementTypes, IFormBuilderDragObject, IFormElement} from '../../../_types';

interface IContainerProps extends Partial<IFormBuilderStateAndDispatch> {
    elementData?: IFormElement;
}

const ContainerWrapper = styled.div<{isOver: boolean; isRootContainer: boolean}>`
    background: ${props => (props.isRootContainer ? '#FFF' : '#EEE')};
    padding: ${props => (props.isRootContainer ? 0 : '2em')};
    color: #aaaaaa;
    text-align: center;
    display: flex;
    flex-direction: column;
    width: 100%;
    min-height: ${props => (props.isRootContainer ? 'calc(100vh - 18rem)' : 'auto')};
`;

const EmptyContainerPlaceholder = styled.div`
    border: 3px dashed #ddd;
    border-radius: 5px;
    text-align: center;
    padding: 2em;
    font-weight: bold;
`;

function Container({elementData, state, dispatch}: IContainerProps): JSX.Element {
    const {t} = useTranslation();
    const dropRef = useRef<HTMLDivElement>(null);
    const [{isOver}, drop] = useDrop<IFormBuilderDragObject<IFormElement>, {containerId: string}, {isOver: boolean}>({
        accept: [
            DraggableElementTypes.ATTRIBUTE,
            DraggableElementTypes.FORM_ELEMENT,
            DraggableElementTypes.RESERVE_LAYOUT_ELEMENT
        ],
        drop: monitor => {
            return {containerId: elementData?.id || defaultContainerId};
        },
        collect: monitor => ({
            isOver: monitor.isOver({shallow: true})
        }),
        hover: (item, monitor) => {
            const isOverCurrent = monitor.isOver({shallow: true});

            if (
                !state ||
                !dispatch ||
                !elementData ||
                !isOverCurrent ||
                item.currentPos?.containerId === elementData?.id ||
                item.element.id === elementData.id
            ) {
                return;
            }

            // Determine rectangle on screen
            const hoverBoundingRect = dropRef.current!.getBoundingClientRect();

            // Determine mouse position
            const clientOffset = monitor.getClientOffset();

            // Get pixels to the top
            const hoverClientY = (clientOffset as XYCoord).y - hoverBoundingRect.top;

            const {index} = item;

            // Element has not been added yet
            const threshold = 0.5;
            const elemHeight = hoverBoundingRect.bottom - hoverBoundingRect.top;
            const bottomThreshold = elemHeight * threshold;

            const addAtPos =
                hoverClientY < bottomThreshold ? 0 : (state.activeElements[elementData.id] ?? []).length || 0;

            const position = {order: addAtPos, containerId: elementData.id};

            if (index < 0) {
                dispatch({
                    type: FormBuilderActionTypes.ADD_ELEMENT,
                    element: {...item.element},
                    position
                });
            } else {
                dispatch({
                    type: FormBuilderActionTypes.MOVE_ELEMENT,
                    elementId: item.element.id,
                    from: item.currentPos ?? {order: item.element.order, containerId: item.element.containerId},
                    to: position
                });
            }
            item.dropAtPos = position;
            item.currentPos = position;
            item.index = position.order;
        }
    });

    if (!state || !dispatch || !elementData) {
        return <></>;
    }

    drop(dropRef);

    const content = state.activeElements[elementData.id] || [];

    return (
        <ContainerWrapper isOver={isOver} isRootContainer={elementData.id === defaultContainerId} ref={dropRef}>
            {!!content.length ? (
                content.map((c, i) => (
                    <FormElementWrapper key={c.id} element={c} index={i} state={state} dispatch={dispatch} />
                ))
            ) : (
                <EmptyContainerPlaceholder>{t('forms.empty_container')}</EmptyContainerPlaceholder>
            )}
        </ContainerWrapper>
    );
}

export default Container;
