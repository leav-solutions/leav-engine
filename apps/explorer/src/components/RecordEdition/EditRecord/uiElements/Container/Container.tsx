// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Col, Row} from 'antd';
import React from 'react';
import {useRecordEditionContext} from '../../hooks/useRecordEditionContext';
import {IFormElementProps} from '../../_types';

function Container({element, record, onValueSubmit, onValueDelete}: IFormElementProps<{}>): JSX.Element {
    const {elements: formElements} = useRecordEditionContext();
    const children = formElements[element.id] ?? [];

    return (
        <>
            {children.map(el => (
                <Row data-testid="container-child-element" key={el.id}>
                    <Col span={24}>
                        {el.uiElement && (
                            <el.uiElement
                                element={el}
                                record={record}
                                onValueSubmit={onValueSubmit}
                                onValueDelete={onValueDelete}
                            />
                        )}
                    </Col>
                </Row>
            ))}
        </>
    );
}

export default Container;
