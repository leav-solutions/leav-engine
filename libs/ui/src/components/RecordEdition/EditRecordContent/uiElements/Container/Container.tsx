// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Col, Row} from 'antd';
import {useRecordEditionContext} from '../../hooks/useRecordEditionContext';
import {IFormElementProps} from '../../_types';

function Container({
    element,
    onValueSubmit,
    onValueDelete,
    onDeleteMultipleValues
}: IFormElementProps<{}>): JSX.Element {
    const {elements: formElements} = useRecordEditionContext();
    const children = formElements[element.id] ?? [];

    return (
        <>
            {children.map(el => (
                <Row data-testid="container-child-element" key={el.id}>
                    <Col span={24}>
                        {el.uiElement && (
                            <el.uiElement
                                element={el}
                                onValueSubmit={onValueSubmit}
                                onValueDelete={onValueDelete}
                                onDeleteMultipleValues={onDeleteMultipleValues}
                            />
                        )}
                    </Col>
                </Row>
            ))}
        </>
    );
}

export default Container;
