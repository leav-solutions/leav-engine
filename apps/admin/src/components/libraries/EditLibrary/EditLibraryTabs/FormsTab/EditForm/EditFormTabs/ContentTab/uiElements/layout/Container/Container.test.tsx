// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {render} from 'enzyme';
import React from 'react';
import {DndProvider} from 'react-dnd';
import {TestBackend} from 'react-dnd-test-backend';
import Container from './Container';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useLayoutEffect: jest.requireActual('react').useEffect
}));

describe('Container', () => {
    test('Snapshot test', async () => {
        const comp = render(
            <DndProvider backend={TestBackend}>
                <Container />
            </DndProvider>
        );

        expect(comp).toMatchSnapshot();
    });
});
