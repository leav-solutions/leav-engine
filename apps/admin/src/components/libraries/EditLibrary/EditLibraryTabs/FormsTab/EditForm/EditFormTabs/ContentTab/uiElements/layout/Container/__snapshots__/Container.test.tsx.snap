// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Snapshot test 1`] = `
LoadedCheerio {
  "length": 0,
  "options": {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
